design-history/exploration-reports/2019.06-unixfsv2-spike-01.md

```diff
@@ -0,0 +1,296 @@
+A unixfsv2 spike, using schemas
+===============================
+
+This document is an exploration of what a hypothetical "unixfs v2" might look like,
+using schemas as a design language for describing the data structures.
+
+This is not a final proposal, but rather a lens to explore both unixfsv2
+concepts and schemas themselves at the same time, and how we expect them
+to contribute to each other.  Several concepts in schemas, advanced layouts,
+their interactions, and their syntaxes will be trialled here; they are not final.
+
+Without further ado...
+
+Schema
+------
+
+```
+type AnyFile union {
+	| File "f"
+	| Dir "d"
+	| Symlink "s"
+} representation keyed
+
+type Filename string
```

Comment:
In the latest round of iterations I’m holding all file and directory metadata in a single property called “meta.” This includes the filename, since it’s also just metadata.

As far as we’re concerned, attributes from a given filesystem (uid, guid, permissions) are no different than the filename or any other custom metadata people might decide to add.

The trick will be to write a strict schema for all the metadata attributes we want to have some semantic meaning around and then make sure IPFS:

1. **Only** ads attributes we’ve defined in the schema.
2. Consistently and non-configurably adds them every time.